lua/packer/compile.lua

```diff
@@ -7,18 +7,57 @@ local luarocks = require('packer.luarocks')
 local config
 local function cfg(_config) config = _config.profile end
 
-local feature_guard = [[
-if !has('nvim-0.5')
-  echohl WarningMsg
-  echom "Invalid Neovim version for packer.nvim!"
-  echohl None
-  finish
-endif
+-- Get function source code if possible
+local files_cache = {}
+
+--  Reads the file_path and caches it's contents for future use
+local function cache_file(file_path)
+  local stat = vim.loop.fs_stat(file_path)
+  if not stat then return false end
+  if not files_cache[file_path] or files_cache[file_path].last_mtime < stat.mtime.sec then
+    files_cache[file_path] = {
+      data = {},
+      last_mtime=vim.loop.fs_stat(file_path).mtime.sec
+    }
+    local file = io.open(file_path, 'r')
+    local file_data = file:read('*a')
+    files_cache[file_path].data = vim.split(file_data, '\n')
+    file:close()
+  end
+  return true
+end
 
-packadd packer.nvim
+-- Returns table containing lines from start to end in file_path
+local function fget_lines(file_path, start, stop)
+  if not cache_file(file_path) then error('file '.. tostring(file_path) .. ' not found') end
+  if start < 1 or stop > #files_cache[file_path].data then
+    error(fmt('Range %d-%d in file %s is out of bounds', start, stop, file_path))
+  end
+  local lines = vim.list_slice(files_cache[file_path].data, start, stop)
+  return lines
+end
+
+-- Returns funcs source of string.dump value
+local function get_function_source(func)
+  if type(func) ~= 'function' then return '' end
+  local src_data = debug.getinfo(func, 'S')
+  if not src_data or not src_data.short_src
+    or not src_data.linedefined or not src_data.lastlinedefined then
+    return vim.inspect(string.dump(func))
+  end
+  local ok, src = pcall(fget_lines, src_data.source:sub(2), src_data.linedefined, src_data.lastlinedefined)
+  if ok and src then
+    local len = #src
+    src[1] = src[1]:sub(src[1]:find('function(.*)'))
+    src[len] = src[len]:sub(1,src[len]:find('end') + 2)
+    return table.concat(src, '\n')
+  end
+  if not ok then vim.api.nvim_err_writeln(src) end
+  return vim.inspect(string.dump(func))
+end
 
-try
-]]
+
+local feature_guard = [[if !has('nvim-0.5') | echohl WarningMsg | echom "Invalid Neovim version for packer.nvim!" | echohl None | finish | endif | packadd packer.nvim | try]]
```

Comment:
nit: is making this one line necessary, it's less readable both in the code and in the compiled file

Reply:
Well the error reports are given based on it possition on where it's in the lua block . For example if the error occurs in line 234 in the compiled file then nvim will report error occured in line 220 . Because current implementation addes 14 lines on top and nvim is coumting lines from `lua << END`. With this nvim will report error in line 220 when error occurs in line line 221 of file . 

This adds way more value IMO .